### PR TITLE
ICU-20941 NumberFormatter: format arbitrary compound units, with inflections

### PR DESCRIPTION
**Pull request by hugovdm**
Closes: icu-units#115, closes: icu-units#27

 - [x] Fix French plural handling (demonstrate with unit tests)
 - [ ] Investigate and fix remaining critical TODOs (check why breaking down built-ins gives different results)

ALLOW_MANY_COMMITS=true

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20941
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/i18n/number_longnames.cpp [hunk 1]
```diff
@@ -119,13 +131,227 @@ static UnicodeString getWithPlural(
...
+
+    UErrorCode localStatus = status;
+    ures_getAllItemsWithFallback(unitsBundle.getAlias(), key.data(), sink, status);
+    if (width == UNUM_UNIT_WIDTH_SHORT) {
+        status = localStatus;
+        return;
+    }
+
+    // Fallback to short. (Apparently this somehow happens automatically in
+    // Java? TODO: find out how!)
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:344]: The fact that it "works" in Java is an unintentional and not robust: https://unicode-org.atlassian.net/browse/ICU-13353

File: icu4c/source/i18n/number_longnames.cpp [hunk 2]
```diff
@@ -430,153 +688,501 @@ const UChar *trimSpaceChars(const UChar *s, int32_t &length) {
...
+        //      singlePluralCategory, singleCaseVariant), such as "{0} metrem"
+        UnicodeString singleUnitArray[ARRAY_LENGTH];
+        // At this point we are left with a Simple Unit:
+        U_ASSERT(uprv_strcmp(singleUnit->build(status).getIdentifier(), singleUnit->getSimpleUnitID()) ==
+                 0);
+        getMeasureData(loc, singleUnit->build(status), width, singleCaseVariant, singleUnitArray,
+                       status);
+        if (U_FAILURE(status)) {
+            // Shouldn't happen if we have data for all single units
+            status = U_UNSUPPORTED_ERROR;
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.cpp:1056]: `status` could be something else like `U_MEMORY_ALLOCATION_ERROR`; you shouldn't overwrite that.

File: icu4c/source/i18n/number_longnames.h [hunk 3]
```diff
@@ -102,18 +102,33 @@ class LongNameHandler : public MicroPropsGenerator, public ModifierStore, public
...
+    // https://unicode.org/reports/tr35/tr35-general.html#compound-units
+    //
+    // TODO: doesn't implement everything yet, e.g. doesn't do times patterns!
+    // This temporarily makes it a bit of a misnomer.
+    //
+    // productUnit is an rvalue reference because this function consumes it,
+    // leaving it in a not-useful / undefined state.
+    // FIXME(review): reasonable enough use of rvalue references? Intended to
+    // indicate that we're taking over the instance in question and that
+    // callers can no longer trust it.
```
Comment (sffc) [inline icu4c/source/i18n/number_longnames.h:125]: Seems fine to me.  C++ is quirky.

**Review by sffc — COMMENTED** [review]
Looks plausible!  I reviewed the general architecture and C++ style.

File: icu4c/source/i18n/number_longnames.cpp [hunk 4]
```diff
@@ -119,13 +131,227 @@ static UnicodeString getWithPlural(
...
 }
 
+enum PlaceholderPosition { PH_EMPTY, PH_NONE, PH_BEGINNING, PH_MIDDLE, PH_END };
+
+// Corresponds to Extract(...) in the spec:
+// https://unicode.org/reports/tr35/tr35-general.html#compound-units
+//
+// FIXME(review): could return a UnicodeString. Stylistic choice: two output
+// parameters kept together as reference parameters, rather than splitting
+// between return value and ref parameter? Idiomatic enough for ICU4C?
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:141]: Done. (with next commit push.)

File: icu4c/source/i18n/number_longnames.cpp [hunk 5]
```diff
@@ -119,13 +131,227 @@ static UnicodeString getWithPlural(
...
+void extractCorePattern(const UnicodeString &pattern,
+                        UnicodeString &coreUnit,
+                        PlaceholderPosition &placeholderPosition,
+                        UChar &joinerChar) {
+    joinerChar = 0;
+    if (pattern.startsWith(u"{0}", 3)) {
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:147]: It looked like it doesn't have the capabilities I need. Worth adding to ICU-20406 yes! (When this PR is considered to be good.)

[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:147]: Dupe.

File: icu4c/source/i18n/number_longnames.cpp [hunk 6]
```diff
@@ -119,13 +131,227 @@ static UnicodeString getWithPlural(
...
+        placeholderPosition = PH_BEGINNING;
+        if (u_isJavaSpaceChar(pattern[3])) {
+            joinerChar = pattern[3];
+            coreUnit.setTo(pattern, 4, pattern.length() - 4);
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:151]: I think you're thinking of unicode surrogate pairs, for example? I think that's not a problem. And operator[] looks to be safe, the offsets are bounds-checked (I checked), and this is documented: "If the offset is not valid (0..length()-1) then U+ffff is returned."

File: icu4c/source/i18n/number_longnames.cpp (hunk 1 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:344]: Adjusted the comment, added a TODO(ICU-13353), thanks!

File: icu4c/source/i18n/number_longnames.cpp (hunk 2 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:1094]: Done.

File: icu4c/source/i18n/number_longnames.h (hunk 3 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.h:125]: Done.

**Comment by hugovdm** [comment]
Managed to trigger a segmentation fault again - digging.

File: icu4c/source/i18n/number_longnames.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp:153]: Can you add a comment to that ticket describing what functionality is missing?

File: icu4c/source/i18n/number_longnames.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (sffc) [reply icu4c/source/i18n/number_longnames.cpp:157]: Cool, didn't know UnicodeString did bounds checking.

File: icu4c/source/common/restrace.cpp [hunk 7]
```diff
@@ -54,6 +54,8 @@ void ResourceTracer::traceOpen() const {
 
 CharString& ResourceTracer::getFilePath(CharString& output, UErrorCode& status) const {
     if (fResB) {
+        // If you get a segfault around here, check that ResourceTable instances
```
Comment (sffc) [inline icu4c/source/common/restrace.cpp:57]: ResourceTable or ResourceArray

**Comment by sffc** [comment]
LGTM, please squash to 2 commits as discussed (after fixing the small typo noted above)

**Comment by hugovdm** [comment]
(Force-pushed twice: I initially had the wrong ticket number.)

File: icu4c/source/i18n/number_longnames.cpp [hunk 8]
```diff
@@ -206,21 +457,26 @@ void getMeasureData(const Locale &locale,
...
         caseKey.append(key, status);
         caseKey.append("/case/", status);
         caseKey.append(unitDisplayCase, status);
 
         UErrorCode localStatus = U_ZERO_ERROR;
+        // TODO(icu-units#138): our fallback logic is not spec-compliant:
+        // lateral fallback should happen before locale fallback. Switch to
+        // getInflectedMeasureData after homogenizing data format? Find a unit
+        // test case that demonstrates the incorrect fallback logic (via
+        // regional variant of an inflected language?)
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:471]: optional:  could you move the description and have only the ticket number.

File: icu4c/source/i18n/number_longnames.cpp [hunk 9]
```diff
@@ -206,21 +457,26 @@ void getMeasureData(const Locale &locale,
...
-        // states we should first look for case="nominative". As part of #138,
-        // either get the spec changed, or add unit tests that warn us if
-        // case="nominative" data differs from no-case data?
     }
 
+    // TODO(icu-units#138): our fallback logic is not spec-compliant: we
+    // check the given case, then go straight to the no-case data. The spec
+    // states we should first look for case="nominative". As part of #138,
+    // either get the spec changed, or add unit tests that warn us if
+    // case="nominative" data differs from no-case data?
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:479]: ditto

File: icu4c/source/i18n/number_longnames.cpp [hunk 10]
```diff
@@ -344,38 +620,59 @@ class DerivedComponents {
...
+                value1_.appendInvariantChars(val1, status);
             }
         }
     }
-    // The returned StringPiece is only valid as long as both the instance
-    // exists, and the compoundValue passed to the constructor is valid.
-    StringPiece value0() const {
-        return sp0;
+
+    // The returned StringPiece is only valid as long as the instance exists.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:637]: optional:
`// Returns a StringPiece that is only valid.... `
or
`// NOTE: the returned .....`

File: icu4c/source/i18n/number_longnames.cpp [hunk 11]
```diff
@@ -344,38 +620,59 @@ class DerivedComponents {
...
+    StringPiece value0(const StringPiece compoundValue) const {
+        return compound0_ ? compoundValue : value0_.toStringPiece();
+    }
+
+    // The returned StringPiece is only valid as long as the instance exists.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:642]: ditto

File: icu4c/source/i18n/number_longnames.cpp [hunk 12]
```diff
@@ -344,38 +620,59 @@ class DerivedComponents {
...
+    StringPiece value1(const StringPiece compoundValue) const {
+        return compound1_ ? compoundValue : value1_.toStringPiece();
+    }
+
+    // The returned char* is only valid as long as the instance exists.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:647]: ditto.

File: icu4c/source/i18n/number_longnames.cpp [hunk 13]
```diff
@@ -344,38 +620,59 @@ class DerivedComponents {
...
+    const char *value0(const char *compoundValue) const {
+        return compound0_ ? compoundValue : value0_.data();
     }
-    // The returned StringPiece is only valid as long as both the instance
-    // exists, and the compoundValue passed to the constructor is valid.
-    StringPiece value1() const {
-        return sp1;
+
+    // The returned char* is only valid as long as the instance exists.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:652]: ditto

**Review by younies — DISMISSED** [review]

File: icu4c/source/common/restrace.cpp [hunk 14]
```diff
@@ -54,6 +54,9 @@ void ResourceTracer::traceOpen() const {
 
 CharString& ResourceTracer::getFilePath(CharString& output, UErrorCode& status) const {
     if (fResB) {
+        // Note: if you get a segfault around here, check that ResourceTable and
+        // ResourceArray instances outlive ResourceValue instances referring to
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/common/restrace.cpp:58]: I would be happy to hear how it's confusing. Given it's optional, I'm going to leave it as is. Definition of "outlive":

**outlive**
verb
(of a person) live longer than (another person).
"women generally outlive men"
survive or last beyond (a specified period or expected lifespan).
"the organization had largely outlived its usefulness"

File: icu4c/source/i18n/measunit_impl.h [hunk 15]
```diff
@@ -310,6 +316,9 @@ class U_I18N_API MeasureUnitImpl : public UMemory {
      * Normalizes a MeasureUnitImpl and generate the identifier string in place.
      */
     void serialize(UErrorCode &status);
+
+    // For calling serialize
+    friend class number::impl::LongNameHandler;
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/measunit_impl.h:321]: Good feedback - I'm making it a TODO though, I filed icu-units#147 - there's more that's wrong with MeasureUnitImpl's serialization. I need time to consider it and fix it properly.

File: icu4c/source/test/intltest/numbertest_api.cpp [hunk 16]
```diff
@@ -1088,6 +1087,167 @@ void NumberFormatterApiTest::unitCompoundMeasure() {
...
+            NumberFormatter::with()
+                .unit(MeasureUnit::forIdentifier("kibijoule-foot-per-cubic-gigafurlong-square-second",
+                                                 status))
+                .unitWidth(UNUM_UNIT_WIDTH_FULL_NAME),
+            Locale("de-CH"),
+            2.4,
+            u"2.4\u00A0Kibijoule⋅Fuss pro Kubikgigafurlong⋅Quadratsekunde");
+
+    // We want to be able to format this, but "100-kilometer" is not yet
+    // supported when it's not part of liter-per-100-kilometer:
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/test/intltest/numbertest_api.cpp:1242]: I filed ICU-21504, thanks for the nudge!

File: icu4c/source/i18n/number_longnames.cpp [hunk 17]
```diff
@@ -119,13 +132,253 @@ static UnicodeString getWithPlural(
...
+    } else {
+        // TODO(icu-units#28): "$unitRes/gender" does not exist. Do we want to
+        // check whether the parent "$unitRes" exists? Then we could return
+        // U_MISSING_RESOURCE_ERROR for incorrect usage (e.g. builtinUnit not
+        // being a builtin).
+        return "";
+    }
+}
+
+// Loads data from a resource tree with paths matching
```
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:226]: Understood. I've mostly been using "//" when I want to add some comments, but not document a function thoroughly yet.

I also don't think it's possible to get Doxygen to actually output documentation for internal functions - but I'd love to be proven wrong, or shown other ways in which more formal documentation can have impact! (Yearning for rustdoc.)

File: icu4c/source/i18n/number_longnames.cpp (hunk 10 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:637]: Done.

File: icu4c/source/i18n/number_longnames.cpp (hunk 11 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:642]: Done.

File: icu4c/source/i18n/number_longnames.cpp (hunk 12 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:647]: Done.

File: icu4c/source/i18n/number_longnames.cpp (hunk 13 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:652]: Done.

**Comment by hugovdm** [comment]
@sffc I'm going ahead and squashing immediately. There were two commits since your last review: 32dc357 and 1f97bfb

**Comment by hugovdm** [comment]
At the previous squashing (following @sffc's review) - 6c64e0f - all tests passed except appveyor, which was still running.

File: icu4c/source/i18n/number_longnames.cpp (hunk 8 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:471]: Leaving as is for now - will consider for the future.

File: icu4c/source/i18n/number_longnames.cpp (hunk 9 above)
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:479]: Leaving as is for now - will consider for the future.

**Review by sffc — APPROVED** [review]

File: icu4c/source/common/restrace.cpp (hunk 7 above)
Comment (hugovdm) [reply icu4c/source/common/restrace.cpp]: Done.

File: icu4c/source/i18n/number_longnames.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/number_longnames.cpp:153]: Done.

**Comment by jefgen** [comment]
Just a heads up, it looks like there are some Valgrind failures on master

```
==19744== Conditional jump or move depends on uninitialised value(s)
==19744==    at 0x528D5B2: value0 (number_longnames.cpp:649)
==19744==    by 0x528D5B2: icu::number::impl::LongNameHandler::forArbitraryUnit(icu::Locale const&, icu::MeasureUnit const&, UNumberUnitWidth const&, char const*, icu::number::impl::LongNameHandler*, UErrorCode&) (number_longnames.cpp:868)
...

==19744== Conditional jump or move depends on uninitialised value(s)
==19744==    at 0x528EDBC: value1 (number_longnames.cpp:654)
==19744==    by 0x528EDBC: icu::number::impl::LongNameHandler::processPatternTimes(icu::MeasureUnitImpl&&, icu::Locale, UNumberUnitWidth const&, char const*, icu::UnicodeString*, UErrorCode&) (number_longnames.cpp:1041)
...
```

I think is the issue is that there is no default value set for `compound0_` and `compound1_`.

```c++
    // Holds strings referred to by value0 and value1;
    bool compound0_, compound1_;
    CharString value0_, value1_;
```

It seems like it should perhaps be set to `false`, but, I'm not 100% sure off-hand what the value should be though.

**Comment by hugovdm** [comment]
PR: https://github.com/unicode-org/icu/pull/1599

Found something similar in Java, fixing in my current PR, hoping it can wait until tomorrow.

**Comment by jefgen** [comment]
Thank you @hugovdm! 👍

**Pull request merged**